### PR TITLE
Avoid chainerx implicit type conversions

### LINKED ISSUE #7515: Forbid chainerx/cupy array mix in FunctionNode and _falblack_workarounds impls
Don't allow numpy cupy arrays mixture in function_node

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes #7515 Needs ~#7531~ ~#7687~ ~#7698~

This prevents to do implicit chainerx conversions when all the inputs to a function are not chainerx variables.

If the implementation is correct I will add tests.

### CODE REVIEW

**Comment by emcastillo** [comment]
There are some issues with the spectal normalization tests

**Comment by niboshi** [comment]
Please address those issues first.

File: tests/chainer_tests/test_function_node.py [hunk 1]
```diff
@@ -362,11 +362,9 @@ def check_mix_xp(self, xp):
         xp_x1 = xp.random.randn(2, 3).astype(numpy.float32)
         xp_x2 = xp.random.randn(2, 3).astype(numpy.float32)
         x2 = backend.to_chx(xp_x2)
-        y, = self.SimpleFunctionNode(xp).apply((xp_x1, x2))
 
-        assert isinstance(y.array, chainerx.ndarray)
-        chainerx.testing.assert_array_equal(
-            backend.CpuDevice().send(xp_x1 * xp_x2), y.array)
+        with self.assertRaises(TypeError):
+            y, = self.SimpleFunctionNode(xp).apply((xp_x1, x2))
```
Comment (niboshi) [inline tests/chainer_tests/test_function_node.py]: * Move instantiation of the function outside the scope.
* No need to assign the output to a variable.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Unrelated Jenkins failure, `DecorrelatedBatchNormalization`

**Comment by niboshi** [comment]
OK. LGTM

**Pull request merged**